File: jelly-roll/jellyroll.py
# -*- coding: utf-8 -*-
"""
Created on Tue May 10 18:58:26 2022

@author: Alex Boisvert

Helper lookup for making "jelly roll" puzzles
"""

import itertools
import os
from collections import defaultdict
import json
import zipfile

# The smallest length for words in the puzzle
MIN_WORD_LENGTH = 4
# The minimum overlap of words
MIN_OVERLAP = 1
# Minimum score of word list entries
MIN_SCORE = 50
# The word list to use
WORDLIST_DIR = os.path.join('..', 'word_lists')
word_list = 'spreadthewordlist.dict'
WORDLIST = os.path.join(WORDLIST_DIR, word_list)

#%% Helper functions

# Make partitions of a string
def multiSlice(s, cutpoints):
    """
    Helper function for allPartitions
    """
    k = len(cutpoints)
    if k == 0:
        return [s]
    else:
        multislices = [s[:cutpoints[0]]]
        multislices.extend(s[cutpoints[i]:cutpoints[i+1]] for i in range(k-1))
        multislices.append(s[cutpoints[k-1]:])
        return multislices

# This includes partitions of length 0
def allPartitions(s, num=None):
    n = len(s)
    cuts = list(range(0, n+1))
    if num:
        num_arr = [num-1]
    else:
        num_arr = range(n)
    for k in num_arr:
        for cutpoints in itertools.combinations_with_replacement(cuts, k):
            yield multiSlice(s, cutpoints)
            
# take pairs of letters from a word
def pairs_of_letters(word_orig):
    # The "keys" of ret are the offset of the word
    # we're interested in 0, 1, 2
    ret = ['', '', '']
    for offset in range(len(ret)):
        remainder = ''
        word = word_orig
        addition, word = word[:2 - offset], word[2 - offset:]
        remainder += addition
        while word:
            addition, word = word[2:4], word[4:]
            remainder += addition
        ret[offset] = remainder
    return ret
            
#%% Read in word list
ALL_WORDS = set()
beginnings = set()
ends = set()

# dictionary of word -> pairs_of_letters
offset_starts = dict()
# dictionary of "finishers" to word beginnings
begin_end_dict = defaultdict(set)

with open(WORDLIST, 'r') as fid:
    for line in fid:
        word, score = line.upper().split(';')
        score = int(score)
        if score >= MIN_SCORE and len(word) >= MIN_WORD_LENGTH:
            ALL_WORDS.add(word)
            # Partition the word to take the beginning and end parts
            for n in range(MIN_OVERLAP, len(word) - MIN_OVERLAP + 1):
                w1, w2 = word[:n], word[n:]
                beginnings.add(w1)
                ends.add(w2)
                
                # we need to find words that finish certain "beginnings"
                # they don't have to "finish" it, just continue it
                for n2 in range(MIN_WORD_LENGTH//2, MIN_WORD_LENGTH+2):
                    w3 = w2[:n2]
                    begin_end_dict[w1].add(w3)
                
            offset_starts[word] = pairs_of_letters(word)

begin_keys = frozenset(begin_end_dict.keys())
            
#%% Make word partitions
# for each word we need to take all partitions of the odd and even
# and we need to see which "beginnings" they can end
# and of course we need to ensure that their "endings" are valid "beginnings"
prev_word_count = 1e6
new_word_count = 0
good_words = ALL_WORDS.copy()

changed_values = (True, True, True)

while max(changed_values):
    gw = set()
    begin_dict = dict()
    end_dict = dict()
    begin_pair_arr = [defaultdict(set), defaultdict(set), defaultdict(set)]
    for word in good_words:
        # see if a partition of the one_two results in endings and beginnings
        os = offset_starts[word]
        # we can keep this word if there's a partition of both that's good
        good_ct = [False, False, False]
        this_begin_dict = defaultdict(set)
        this_end_dict = defaultdict(set)
        for i, w in enumerate(os):
            for n in range(MIN_OVERLAP, len(w) - MIN_OVERLAP + 1):
                w1, w2 = w[:n], w[n:]
                if w2 in beginnings and w1 in ends:
                    this_word = word
                    this_begin_dict[w1] |= set([this_word])
                    this_end_dict[w2] |= set([this_word])
                    begin_pair_arr[i][w1] |= set([this_word])
                    good_ct[i] = True
        if min(good_ct):
            for k, v in this_begin_dict.items():
                begin_dict[k] = begin_dict.get(k, set()).union(v)
            for k, v in this_end_dict.items():
                end_dict[k] = end_dict.get(k, set()).union(v)
            gw.add(word)
            
    print(len(good_words))
    orig_lengths = (len(good_words), len(beginnings), len(ends))
    good_words = good_words.intersection(gw)
    print(len(good_words))
    beginnings = beginnings.intersection(set(begin_dict.keys()))
    ends = ends.intersection(set(end_dict.keys()))
    new_lengths = (len(good_words), len(beginnings), len(ends))
    changed_values = [orig_lengths[i] == new_lengths[i] for i in range(len(orig_lengths))]
    
# Remove blank entries from begin_pair_arr
tmp = [{}, {}, {}]
begin_pair_arr2 = [{}, {}, {}]
for i, d in enumerate(begin_pair_arr):
    for k, v in d.items():
        if v:
            tmp[i][k] = v
            begin_pair_arr2[i][k] = list(v)
begin_pair_arr = tmp

#%% Write file to zipped JSON for JS purposes
# Note that we can create "begin_keys" from begin_end_dict in JS

begin_end_dict2 = dict((k, list(v)) for k, v in begin_end_dict.items())

j = {
  "good_words": list(good_words)
, "begin_end_dict": begin_end_dict2
, "begin_pair_arr": begin_pair_arr2
}

# Convert the Python object to a JSON string
json_data = json.dumps(j)

# Create a new ZIP file and add the JSON data to it
with zipfile.ZipFile("jellyroll.json.zip", 'w', zipfile.ZIP_DEFLATED) as zip_file:
    # Write JSON data to a file inside the ZIP
    zip_file.writestr("jellyroll.json", json_data)

#%%
def add_word(word, all_words, white_words, gray_words):
    """
    Add a word to all_words. This updates one and two words too.
    """
    all_words2 = all_words + [word]
    all_string = ''.join(all_words2)
    white_string = ''.join(white_words)
    gray_string = ''.join(gray_words)
    
    # We have to pull out any "one" or "two" words 
    # that resulted from adding our word
    white_start = pairs_of_letters(all_string)[1][len(white_string):]
    gray_start = pairs_of_letters(all_string[1:])[0][len(gray_string):]
    
    white_words2, gray_words2 = white_words, gray_words
    
    if begin_end_dict.get(white_start):
        # it's okay if there are future words we can make from this
        pass
    else:
        for i in range(1, len(white_start)):
            if white_start[:-i] in good_words:
                white_words2 = white_words2 + [white_start[:-i]]
                white_start = white_start[-i:]
                break
    if begin_end_dict.get(gray_start):
        pass
    else:
        for i in range(1, len(gray_start)):
            if gray_start[:-i] in good_words:
                gray_words2 = gray_words2 + [gray_start[:-i]]
                gray_start = gray_start[-i:]
                break
    return all_words2, white_words2, gray_words2, white_start, gray_start
    
def does_word_work(word, all_words, white_words, gray_words):
    """
    Check if a word works with our current words
    """
    # Add the word to our list
    all_words2, white_words2, gray_words2, white_start, gray_start = add_word(word, all_words, white_words, gray_words)
    all_string = ''.join(all_words2)

    # We start with white if the length mod 4 is 0 or 3
    start_white = False
    if len(all_string) % 4 in (0, 3):
        start_white = True
        
    # We start with one letter if the length is even
    start_one = False
    if len(all_string) % 2 == 0:
        start_one = True
    
    # find a new "all" word that "ends" both the words
    white_endings = begin_end_dict[white_start]
    gray_endings = begin_end_dict[gray_start]
    next_white = set()
    next_gray = set()

    if not start_white and not start_one:
        for x in gray_endings:
            next_gray |= begin_pair_arr[0].get(x, set())
        for x in white_endings:
            next_white |= begin_pair_arr[2].get(x, set())
    elif start_white and not start_one:
        for x in gray_endings:
            next_gray |= begin_pair_arr[2].get(x, set())
        for x in white_endings:
            next_white |= begin_pair_arr[0].get(x, set())
    elif start_white and start_one:
        for x in gray_endings:
            next_gray |= begin_pair_arr[2].get(x, set())
        for x in white_endings:
            next_white |= begin_pair_arr[1].get(x, set())
    elif not start_white and start_one:
        for x in gray_endings:
            next_gray |= begin_pair_arr[1].get(x, set())
        for x in white_endings:
            next_white |= begin_pair_arr[2].get(x, set())
            
    possible_next_words = next_gray & next_white
    
    # Instead of doing a recursive check, just ensure
    # that the "ends" of these are valid "beginnings"
    ret = set()
    for pnw in possible_next_words:
        _, _, _, white_start, gray_start = add_word(pnw, all_words2, white_words2, gray_words2)
        if white_start in begin_keys and gray_start in begin_keys:
            ret.add(pnw)
    return ret

def next_inner_words(word, all_words, white_words, gray_words):
    """
    Print the next "inner" words that would arise if we added "word"
    """
    _, white_words2, gray_words2, white_start, gray_start = add_word(word, all_words, white_words, gray_words)
    return [' '.join(white_words2[len(white_words):]), ' '.join(gray_words2[len(gray_words):]), white_start, gray_start]
    
def next_word_sorter(next_word, all_words, white_words, gray_words):
    """
    Sort "next words" by length descending
    """
    next_white, next_gray, white_start, gray_start = next_inner_words(next_word, all_words, white_words, gray_words)
    white_len, gray_len = len(''.join(next_white)), len(''.join(next_gray))
    if not white_len:
        white_len = len(white_start) + 2
    if not gray_len:
        gray_len = len(gray_start) + 2
    return white_len + gray_len + len(next_word)
    
#%%
all_words, white_words, gray_words = [], [], []
word = 'runinto'

word = word.upper()

while True:
    # Print our possibles
    next_words = does_word_work(word, all_words, white_words, gray_words)
    if next_words:
        all_words, white_words, gray_words, white_start, gray_start = add_word(word, all_words, white_words, gray_words)
        # Print our current length
        print(len(''.join(all_words)))
        print(all_words, white_words, gray_words)
        print(white_start, gray_start)
        next_words = sorted(next_words, key=lambda x: next_word_sorter(x, all_words, white_words, gray_words), reverse=True)
        for nw in next_words[:15]:
            niw = next_inner_words(nw, all_words, white_words, gray_words)
            print(' / '.join([nw] + niw))
    else:
        # What do we do in this case?
        print("No further fill found. Backtracking needed.")
        break
    
    # Do a loop for choosing the next word
    word = input("Enter the next word: ").upper()
    remain_in_loop = True
    while remain_in_loop:
        if does_word_work(word, all_words, white_words, gray_words):
            remain_in_loop = False
        else:
            word = input("That word doesn't work. Choose another: ").upper()
            
        
    

File: jelly-roll/jellyroll.js
/**
* Filling logic for jelly roll puzzles
* (c) 2024, Crossword Nexus.
* MIT License - https://opensource.org/license/MIT
**/

// The Begin Button
const BEGIN_BUTTON = document.getElementById('begin-button');

// Helper functions to find the intersection and union of two sets
function setIntersection(set1, set2) {
  return new Set([...set1].filter(x => set2.has(x)));
}

function setUnion(set1, set2) {
  if (!set2) return set1;
  return new Set([...set1, ...set2]);
}

// take pairs of letters from a word
function pairsOfLetters(wordOrig) {
  // The "keys" of ret are the offset of the word
  // we're interested in 0, 1, 2
  let ret = ['', '', ''];
  for (var offset = 0; offset < ret.length; offset++) {
    let remainder = '';
    let word = wordOrig;
    // cut off the beginning to add
    let addition = word.substr(0, 2 - offset);
    word = word.substr(2 - offset);
    remainder += addition;
    while (word.length) {
      addition = word.substr(2, 2);
      word = word.substr(4);
      remainder += addition;
    }
    ret[offset] = remainder
  }
  return ret
}

// Helper function to slice a string like in Python
function sliceString(str, { i = 0, j = str.length, k = 1 } = {}) {
  let result = '';
  // Adjust negative indices for i and j (if provided)
  if (i < 0) i = str.length + i;
  if (j < 0) j = str.length + j;
  // Loop through the string with the given step k
  for (let index = i; index < j && index >= 0 && index < str.length; index += k) {
    result += str[index];
  }
  return result;
}

// add a word to our existing data, updating all arrays
function addWord(word, allWords, whiteWords, grayWords) {
  const allWords2 = allWords.concat([word]);
  const allString = allWords2.join('');
  const whiteString = whiteWords.join('');
  const grayString = grayWords.join('');

  // pull out any white or gray words
  // that result from adding our word
  var whiteStart = pairsOfLetters(allString)[1].slice(whiteString.length);
  var grayStart = pairsOfLetters(allString.substr(1))[0].slice(grayString.length);

  var whiteWords2 = whiteWords;
  var grayWords2 = grayWords;

  if (!BEGIN_END_DICT[whiteStart]) {
    const l1 = whiteStart.length;
    for (var i=1; i < l1; i++) {
      if (GOOD_WORDS.has(whiteStart.substr(0, l1-i))) {
        whiteWords2 = whiteWords2.concat(whiteStart.substr(0, l1-i));
        whiteStart = whiteStart.slice(-i);
        break;
      }
    }
  }
  if (!BEGIN_END_DICT[grayStart]) {
    const l1 = grayStart.length;
    for (var i=1; i<l1; i++) {
      if (GOOD_WORDS.has(grayStart.substr(0, l1-i))) {
        grayWords2 = grayWords2.concat(grayStart.substr(0, l1-i));
        grayStart = grayStart.slice(-i);
        break;
      }
    }
  }
  return [allWords2, whiteWords2, grayWords2, whiteStart, grayStart];
}

// Check if a new word will work with our current word
function doesWordWork(word, allWords, whiteWords, grayWords) {
  // add the word to new arrays
  let [allWords2, whiteWords2, grayWords2, whiteStart, grayStart] = addWord(word, allWords, whiteWords, grayWords);
  const allString = allWords2.join('');

  // We start with white if the length mod 4 is 0 or 3
  const startWhite = [0, 3].includes(allString.length % 4);
  // we start with one letter if the length is even
  const startOne = (allString.length %2 == 0);

  // Find a new "all" word that "ends" the words
  const whiteEndings = BEGIN_END_DICT[whiteStart];
  const grayEndings = BEGIN_END_DICT[grayStart];
  var nextWhite = new Set();
  var nextGray = new Set();

  // no need to go through all this if there's nothing
  if (!whiteEndings || !grayEndings) {
    return new Set();
  }

  if (!startWhite && !startOne) {
    grayEndings.forEach(x => {
      nextGray = setUnion(nextGray, BEGIN_PAIR_ARR[0][x]);
    });
    whiteEndings.forEach(x => {
      nextWhite = setUnion(nextWhite, BEGIN_PAIR_ARR[2][x]);
    });
  } else if (startWhite && !startOne) {
    grayEndings.forEach(x => {
      nextGray = setUnion(nextGray, BEGIN_PAIR_ARR[2][x]);
    });
    whiteEndings.forEach(x => {
      nextWhite = setUnion(nextWhite, BEGIN_PAIR_ARR[0][x]);
    });
  } else if (startWhite && startOne) {
    grayEndings.forEach(x => {
      nextGray = setUnion(nextGray, BEGIN_PAIR_ARR[2][x]);
    });
    whiteEndings.forEach(x => {
      nextWhite = setUnion(nextWhite, BEGIN_PAIR_ARR[1][x]);
    });
  } else if (!startWhite && startOne) {
    grayEndings.forEach(x => {
      nextGray = setUnion(nextGray, BEGIN_PAIR_ARR[1][x]);
    });
    whiteEndings.forEach(x => {
      nextWhite = setUnion(nextWhite, BEGIN_PAIR_ARR[2][x]);
    });
  }

  // The intersection of these is what we want
  const possibleNextWords = setIntersection(nextGray, nextWhite);

  // do a little check to make sure these words can be continued
  var ret = new Set();
  for (var pnw of possibleNextWords) {
    [_, _, _, whiteStart, grayStart] = addWord(pnw, allWords2, whiteWords2, grayWords2);
    if (BEGIN_KEYS.has(whiteStart) && BEGIN_KEYS.has(grayStart)) {
      ret.add(pnw);
    }
  }
  return ret;
} // END doesWordWork()

// find the next "inner" words that would arise if we added "word"
function nextInnerWords(word, allWords, whiteWords, grayWords) {
  [_, whiteWords2, grayWords2, whiteStart, grayStart] = addWord(word, allWords, whiteWords, grayWords);
  return [whiteWords2.slice(whiteWords.length).join(' '), grayWords2.slice(grayWords.length).join(' '), whiteStart, grayStart];
}

// helper function to sort the next words
function nextWordSorter(nextWord, allWords, whiteWords, grayWords) {
  [nextWhite, nextGray, whiteStart, grayStart] = nextInnerWords(nextWord, allWords, whiteWords, grayWords);
  let whiteLen = nextWhite.length;
  let grayLen = nextGray.length;
  if (whiteLen == 0) {
    whiteLen = nextWhite.length + 2;
  }
  if (grayLen == 0) {
    grayLen = nextGray.length + 2;
  }
  return whiteLen + grayLen + nextWord.length;
}

/** Handle button click events **/
function handleClick(newData=[]) {
  // change the button to "loading"
  buttonLoading(BEGIN_BUTTON);

  // Run all this asynchronously
  setTimeout(() => {
    // grab the data from the text boxes
    let allWords1 = document.getElementById('jelly-roll').value.toUpperCase().split('\n').filter(Boolean);
    let whiteWords = document.getElementById('white-squares').value.toUpperCase().split('\n').filter(Boolean);
    let grayWords = document.getElementById('gray-squares').value.toUpperCase().split('\n').filter(Boolean);

    // assume the last word is what we're adding
    let allWords = allWords1.slice(0, -1);
    let word = allWords1[allWords1.length - 1];

    // get our possibles
    let nextWords = doesWordWork(word, allWords, whiteWords, grayWords);

    // if there's nothing here, we create an alert and move on
    if (nextWords.size === 0) {
      alert("No fill found. Try a different entry.")
      // remove the new entry/entries from the text boxes
      const textBoxes = ['jelly-roll', 'white-squares', 'gray-squares'];
      for (i = 0; i < textBoxes.length; i++) {
        if (newData[i]) {
          let arr = document.getElementById(textBoxes[i]).value.toUpperCase().split('\n').filter(Boolean);
          arr.pop();
          document.getElementById(textBoxes[i]).value = arr.join('\n');
        }
      }
    } else {
      // add the word
      var whiteStart, grayStart;
      [allWords, whiteWords, grayWords, whiteStart, grayStart] = addWord(word, allWords, whiteWords, grayWords);

      // sort nextwords
      nextWords = [...nextWords].sort((a, b) =>
        nextWordSorter(b, allWords, whiteWords, grayWords) - nextWordSorter(a, allWords, whiteWords, grayWords)
      );

      //console.log(nextWords);
      // create the data set for the table
      const dataSet = [];
      for (var i=0; i<nextWords.length; i++) {
        const niw = nextInnerWords(nextWords[i], allWords, whiteWords, grayWords);
        const nextRow = [nextWords[i], niw[0], niw[1], niw[2], niw[3]];
        //console.log(niw);
        dataSet.push(nextRow);
      }
      // fill the table
      var table = $('#datatables-table').DataTable();
      table.order([]);
      table.clear().rows.add(dataSet).draw();
      // Clear the search bar
      table.search('').draw();

      // update the length
      document.getElementById('jelly-roll-length').innerHTML = `(${allWords1.join('').length})`;

    } // end if/else

    // Change the button back to "active"
    buttonActive(BEGIN_BUTTON);
  }, 0);
} // end handleClick()

/** Define what happens when we click on a row in the table **/
$('#datatables-table tbody').on('click', 'tr', function() {
    // Grab data from text boxes
    let allWords = document.getElementById('jelly-roll').value.toUpperCase().split('\n').filter(Boolean);
    let whiteWords = document.getElementById('white-squares').value.toUpperCase().split('\n').filter(Boolean);
    let grayWords = document.getElementById('gray-squares').value.toUpperCase().split('\n').filter(Boolean);

    // Grab the data from the clicked row
    const data = table.row(this).data();
    // add relevant data to relevant arrays
    if (data[0]) allWords.push(data[0]);
    if (data[1]) whiteWords.push(data[1]);
    if (data[2]) grayWords.push(data[2]);

    // re-populate text boxes
    document.getElementById('jelly-roll').value = allWords.join('\n');
    document.getElementById('white-squares').value = whiteWords.join('\n');
    document.getElementById('gray-squares').value = grayWords.join('\n');

    // Click the button
    handleClick(newData=data);
});

/**  Change button's functionality and color and text **/

// Change to a "loading" button
function buttonLoading(button) {
  button.textContent = 'Loading ...';
  button.style['background-color'] = 'gray';
  button.style.cursor = 'default';
  button.removeEventListener('click', handleClick);
}

function buttonActive(button) {
  button.style['background-color'] = '#007bff';
  button.innerHTML = 'Begin';
  button.style.cursor = 'pointer';
  button.addEventListener('click', handleClick);
}

/** Export a puzzle in RG format **/
function exportToVpuz(_) {
  // Set up the output variable
  var vpuzOut = {"author": "AUTHOR_GOES_HERE", "title": "TITLE_GOES_HERE", "copyright": "COPYRIGHT_GOES_HERE"};
  const notes = `Three paths curl toward the center of this jelly roll — a white path, a gray path, and a “jelly roll” path that uses all the letters of the other two (winding back and forth as indicated by the heavy bars). Answer words proceed one after the other. Word lengths are given for the answers in the white and gray paths. It’s up to you to determine where the jelly roll answers begin and end.`;
  vpuzOut['notes'] = notes;

  // Read the text boxes
  let allArr = document.getElementById('jelly-roll').value.split('\n');

  // solution string
  const solutionString = allArr.join('').split('').sort().join('');
  vpuzOut['solution-string'] = solutionString;

  // clues
  const clues = {};
  const CLUE_HEADERS = ["White Path", "Gray Path", "Jelly Roll"];
  const ELTS = ['white-squares', 'gray-squares', 'jelly-roll'];
  for (var i=0; i < CLUE_HEADERS.length; i++) {
    let arr = document.getElementById(ELTS[i]).value.split('\n');
    clues[CLUE_HEADERS[i]] = [];
    arr.forEach(x => {
      let clue = `CLUE_FOR_${x}`;
      if (ELTS[i] != 'jelly-roll') clue += ` ${x.length}`;
      clues[CLUE_HEADERS[i]].push(clue);
    });
  }
  vpuzOut["clues"] = clues;

  // image
  const numLetters = solutionString.length / 2; // solution length must be even
  // draw on the canvas
  drawJellyRoll(numLetters);
  // trim
  const trimmedCanvas = getTrimmedCanvas();
  const base64String = trimmedCanvas.toDataURL("image/png");
  vpuzOut['puzzle-image'] = base64String;

  let fileName = allArr[0] + '.vpuz';
  let vpuzString = JSON.stringify(vpuzOut, null, 2);
  downloadStringAsFile(vpuzString, fileName, "text/plain");

}

// Export button functionality
document.getElementById('export-button').addEventListener('click', exportToVpuz);

// Helper function for creating downloads
function downloadStringAsFile(content, fileName, mimeType) {
    // Create a blob with the content
    const blob = new Blob([content], { type: mimeType });

    // Create a temporary anchor element
    const a = document.createElement('a');
    a.href = URL.createObjectURL(blob);
    a.download = fileName;

    // Append the anchor to the body (required for some browsers)
    document.body.appendChild(a);

    // Trigger the download by simulating a click
    a.click();

    // Remove the anchor element from the document
    document.body.removeChild(a);

    // Revoke the object URL to free up memory
    URL.revokeObjectURL(a.href);
}
